vendor/harfbuzzFreetype/harfbuzz_wrap.cpp

```diff
@@ -0,0 +1,108 @@
+    buffer = hb_buffer_create();
+
+    hb_buffer_allocation_successful(buffer);
+}
+
+HBShaperWrap::~HBShaperWrap() {
+    if (!face.valid) return;
+    hb_buffer_destroy(buffer);
+    hb_font_destroy(font);
+}
+
+void HBShaperWrap::CreateComplexGlyphIDs(const std::string &text,
+                                         std::vector<uint32_t>& glyphIDs,
+                                         std::vector<HBShapeAdjust> &adjusts) {
+    // Setup harfbuzz
+    hb_buffer_reset(buffer);
+
+    hb_buffer_set_direction(buffer, hbInfo->direction);
+    hb_buffer_set_script(buffer, hbInfo->script);
+    hb_buffer_set_language(buffer, hb_language_from_string(hbInfo->language.c_str(), (int)hbInfo->language.size()));
+    size_t length = text.size();
+
+    hb_buffer_add_utf8(buffer, text.c_str(), (int)length, 0, (int)length);
+
+    // harfbuzz shaping
+    hb_shape(font, buffer, NULL, 0);
+
+    // Get Harfbuzz adjustion
+    uint32_t glyphCount;
+    hb_glyph_info_t *glyphInfo = hb_buffer_get_glyph_infos(buffer, &glyphCount);
+    hb_glyph_position_t *glyphPos = hb_buffer_get_glyph_positions(buffer, &glyphCount);
+
+    glyphIDs.reserve(glyphCount);
+    adjusts.reserve(glyphCount);
+
+    for (uint32_t i = 0; i < glyphCount; ++i) {
+        glyphIDs.emplace_back(glyphInfo[i].codepoint);
+
+        float x_advance = static_cast<float>(glyphPos[i].x_advance / 64.0f);
+        float x_offset = static_cast<float>(glyphPos[i].x_offset / 64.0f);
+        float y_offset = static_cast<float>(glyphPos[i].y_offset / 64.0f);
+
+        adjusts.emplace_back(x_offset, y_offset, x_advance);
+    }
+}
+
+void HBShaperWrap::CreateComplexGlyphIDs(const std::u16string& text,
+                                         std::vector<uint32_t>& glyphIDs,
+                                         std::vector<HBShapeAdjust>& adjusts) {
+    // Setup harfbuzz
+    hb_buffer_reset(buffer);
+
+    hb_buffer_set_direction(buffer, hbInfo->direction);
+    hb_buffer_set_script(buffer, hbInfo->script);
+    hb_buffer_set_language(buffer, hb_language_from_string(hbInfo->language.c_str(), (int)hbInfo->language.size()));
+    size_t length = text.size();
+
+    hb_buffer_add_utf16(buffer, (const uint16_t *)text.c_str(), (int)length, 0, (int)length);
+
+    // harfbuzz shaping
+    hb_shape(font, buffer, NULL, 0);
+
+    // Get Harfbuzz adjustion
+    uint32_t glyphCount;
+    hb_glyph_info_t *glyphInfo = hb_buffer_get_glyph_infos(buffer, &glyphCount);
+    hb_glyph_position_t *glyphPos = hb_buffer_get_glyph_positions(buffer, &glyphCount);
+
+    glyphIDs.reserve(glyphCount);
+    adjusts.reserve(glyphCount);
+
+    for (uint32_t i = 0; i < glyphCount; ++i) {
+        glyphIDs.emplace_back(glyphInfo[i].codepoint);
+
+        float x_advance = static_cast<float>(glyphPos[i].x_advance / 64.0f);
+        float x_offset = static_cast<float>(glyphPos[i].x_offset / 64.0f);
+        float y_offset = static_cast<float>(-glyphPos[i].y_offset / 64.0f);
```

Comment:
Why the offset is negated here compared to L67 above?

Reply:
Thanks for correcting this, L67 should be negated, will correct it in next commit.

Reply:
origin L67 removed